syntax="proto2";

import "braft/enum.proto";

package braft;
option cc_generic_services = true;

// data store in baidu-rpc's attachment
message EntryMeta {
    required int64 term = 1;
    required EntryType type = 2;
    repeated string peers = 3;
    optional uint64 data_len = 4;
    // Don't change field id of `old_peers' in the consideration of backward
    // compatibility
    repeated string old_peers = 5;
    // RS-Raft
    optional uint32 k = 6;
    optional uint32 fragment_index = 7;
    optional uint32 fragment_len = 8;
    optional uint32 valid_data_len = 9;
    optional uint32 metadata_len = 10;
};

message RequestVoteRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
    required int64 last_log_term = 5;
    required int64 last_log_index = 6;
    optional int64 priority = 7 [default = 0];
};

message RequestVoteResponse {
    required int64 term = 1;
    required bool granted = 2;
    optional int64 priority = 3 [default = 0];
};

message AppendEntriesRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
    required int64 prev_log_term = 5;
    required int64 prev_log_index = 6;
    repeated EntryMeta entries = 7;
    required int64 committed_index = 8;
    optional int64 leader_priority = 9 [default = -1];
};

message AppendEntriesResponse {
    required int64 term = 1;
    required bool success = 2;
    optional int64 last_log_index = 3;
    optional bool readonly = 4;
    optional int64 priority = 5 [default = 0];
};

message SnapshotMeta {
    required int64 last_included_index = 1;
    required int64 last_included_term = 2;
    repeated string peers = 3;
    repeated string old_peers = 4;
}

message InstallSnapshotRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
    required SnapshotMeta meta = 5;
    required string uri = 6;
};

message InstallSnapshotResponse {
    required int64 term = 1;
    required bool success = 2;
};

message TimeoutNowRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
}

message TimeoutNowResponse {
    required int64 term = 1;
    required bool success = 2;
}

// RS-Raft
message FragmentMeta {
    optional uint32 k = 6;
    optional uint32 fragment_index = 7;
    optional uint32 fragment_len = 8;
    optional uint32 valid_data_len = 9;
    optional bool valid = 10;
}

message CollectFragmentRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
    repeated uint32 metadata_len = 5; 
    required int64 first_collect_index = 6;
    required int64 last_collect_index = 7;
}

message CollectFragmentResponse {
    required int64 term = 1;
    required bool success = 2;
    repeated FragmentMeta meta = 3;
    required int64 first_collect_index = 4;
    required int64 last_collect_index = 5;
}

service RaftService {
    rpc pre_vote(RequestVoteRequest) returns (RequestVoteResponse);

    rpc request_vote(RequestVoteRequest) returns (RequestVoteResponse);

    rpc append_entries(AppendEntriesRequest) returns (AppendEntriesResponse);

    rpc install_snapshot(InstallSnapshotRequest) returns (InstallSnapshotResponse);

    rpc collect_fragment(CollectFragmentRequest) returns (CollectFragmentResponse);

    rpc timeout_now(TimeoutNowRequest) returns (TimeoutNowResponse);
};

